### PR TITLE
fix bug in int128

### PR DESCRIPTION
**Pull request by krux02**
fix bug reported by @disruptek 

This bug is a bit embarrassing to me since integer arithmetic is quite important to be correct. Therefore it is important that this fix will be merged.

### CODE REVIEW

**Comment by disruptek** [comment]
No reason to be embarrassed; I'd say you more than made up for it on the test -- wow!

File: compiler/int128.nim [hunk 1]
```diff
@@ -696,3 +696,86 @@ when isMainModule:
   doAssert divMod(toInt128(-mb),toInt128( ma)) == (toInt128(-mb div  ma), toInt128(-mb mod  ma))
   doAssert divMod(toInt128( mb),toInt128(-ma)) == (toInt128( mb div -ma), toInt128( mb mod -ma))
   doAssert divMod(toInt128(-mb),toInt128(-ma)) == (toInt128(-mb div -ma), toInt128(-mb mod -ma))
+
+  let e = parseDecimalInt128("70997106675279150998592376708984375")
+
+  let strArray = [
```
Comment (cooldome) [inline compiler/int128.nim:702]: use const

File: compiler/int128.nim [hunk 2]
```diff
@@ -696,3 +696,86 @@ when isMainModule:
...
+    "00000000000000000000000000000000", "00000000000000000000000000000000", "00000000000000000000000000000000", "00000000000000000000000000000000",
+    "00000000000000000000000000000000", "00000000000000000000000000000000", "00000000000000000000000000000000", "00000000000000000000000000000000",
+  ]
+
+  for i in 0 ..< 128:
+    let str1 = toHex(e shr i)
+    let str2 = strArray[i]
+    doAssert str1 == str2
+
+  let strArray2 = [
```
Comment (cooldome) [inline compiler/int128.nim:743]: use const

**Comment by Araq** [comment]
I ... cannot ... *dies*

**Review by disruptek — APPROVED** [review]

**Review by Clyybber — APPROVED** [review]

**Comment by timotheecour** [comment]
can you provide in PR top message an input that was previously wrong and is now fixed?

**Comment by Araq** [comment]
> This bug is a bit embarrassing to me since integer arithmetic is quite important to be correct. Therefore it is important that this fix will be merged.

Not just to you... Is it really correct **now**? Do we backport the fix and release 1.0.8 just for that? How about we use Status's integer library instead. This is at least the 3rd bug in a serious of bugs that should never have happened.

**Comment by timotheecour** [comment]
semi-related: when int128 was introduced (to fix overflow issues), was there any benchmarking done to ensure that there was no resulting loss in performance (eg compile times, for eg bootstrap time for compiling nim or other code that would contain lots of integers) ?

I would expect 128 bit arithmetic to be at least 2X slower than builtin arithmetics, and that change was (IIRC) applied to all PNode's of kind intVal

**Comment by krux02** [comment]
> Not just to you... Is it really correct now? Do we backport the fix and release 1.0.8 just for that? 

I just did some damage analysis to find out how much of this Bug is affected. And I am a bit relaxed now. ``shr`` on `Int128` was incorrect, if the argument was in the range ``32 ..< 96``. `shr` on `Int128` is used within the `int128` library exclusively, and there it is used only for `divMod` with a constant shift of `1`. Nothing is really affected by this bug. `shr` in semfold casts to `uint64`/`int64` before it does anything. Same in `vm` code. So in other words, no we do not need to backport anything. But just to be sure I will test `int128` a bit more extensively now.

The reason why this bug was detected was beceause @disruptek wanted to use this `int128` for the super advanced floating point printing a bit more extensively and then hit this bug.

**Comment by Araq** [comment]
> I would expect 128 bit arithmetic to be at least 2X slower than builtin arithmetics, and that change was (IIRC) applied to all PNode's of kind intVal

No benchmarking was done for two reasons: 

1. Correctness is more important than compiler speed.
2. I expected it to be off the critical path as we lose most time in node allocations and traversals. We simply don't do much arithmetic in the compiler at all.

**Comment by Araq** [comment]
> I just did some damage analysis to find out how much of this Bug is affected. And I am a bit relaxed now. 

Excellent, thank you!

**Pull request merged**